docs/fornax-design/mission_design.md

```diff
@@ -0,0 +1,90 @@
+
+```golang
+// Mission specifies a workload to deploy in edge clusters
+type Mission struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec MissionSpec `json:"spec"`
+	State map[string]string `json:"state,omitempty"`
+}
+
+type MissionSpec struct {
+	MissionResource string `json:"missionresource,omitempty"`
+	MissionCommand MissionCommandSpec `json:"missioncommand,omitempty"`
+	Placement GenericPlacementFields `json:"placement,omitempty"`
+	StateCheck StateCheckFields `json:"statecheck"`
+}
+
+type MissionCommandSpec struct {
+	Trigger string `json:"trigger,omitempty"`
+	RunIfTriggerSucceed bool `json:"runiftriggersucceed"`
+	Command string `json:"command,omitempty"`
+	ReverseCommand string `json:"reversecommand,omitempty"`
+}
+
+type GenericClusterReference struct {
+	Name string `json:"name"`
+}
+
+type GenericPlacementFields struct {
+	Clusters    []GenericClusterReference `json:"clusters,omitempty"`
+	MatchLabels map[string]string         `json:"matchlabels,omitempty"`
+}
+type StateCheckFields struct {
+	Command string `json:"command,omitempty"`
+}
+```
+
+## Mission/MissionSpec/MssionResource
+This field is the content of a Kubernetes resource in a yaml file. 
+
+Fornax will
+1. use command "kubectl apply -f" to deploy the mission resource, 
+2. use command "kubectl delete -f" to delete the mission resource, 
+3. use command "kubectl get [resource_type] [resource_name]" to grab the state of mission content, if the mission does not specify the mission state check command. 
+
+Note that Fornax prefers MissionResource over MissionCommand. Users are encoraged to use MissionResource to describe the mission content as long as it is possible, and use the MissionCommand only if they have to. We get the most of the nice feature of declarative object management in Kubernetes if we delcare the mission content using MissionResource.
+
+If both MissionResource and MissionCommand fields are specified in a mission, the MissionCommand will be ignored. A mission is invalid if neither issionResource nor MissionCommand field is specified.
+
+## Mission/MissionSpec/MssionCommand
+
+We use MissionCommand if MissionResource is insufficient, for example:
+1. We need to run a kubectl command other than "kkubectl apply" to do the work.
+2. Or, we need to do something via a bash command
+
+The Content of the MssionCommand is defined via multiple fields:
+1. Trigger: the command to check whether or not we need to run the mission command. By default, the mission command will be run only if the trigger command fails.
+2. RunIfTriggerSucceed: by default it is false, which means the mission command will be run only if the trigger command fails. The mission command will run if the trigger command succeeds, if this value is set to true.
```

Comment:
ah i see the definition here. the meaning could be a bit confusing when it's set to false because it's not clear from the name whether false applies to Run or does false apply to "Succeed".

Reply:
OK, I rename RunIfTriggerSucceed to RunWhenTriggerSucceed.